transforms3d/quaternions.py

```diff
@@ -304,10 +304,105 @@ def qinverse(q):
     return qconjugate(q) / qnorm(q)
 
 
-def qeye():
+def qeye(dtype = np.float):
     ''' Return identity quaternion '''
-    return np.array([1.0,0,0,0])
+    return np.array([1.0,0,0,0], dtype = dtype)
 
+def qexp(q):
+    ''' Return exponential of quaternion
+    Parameters
+    ----------
+    q : 4 element sequence
+       w, i, j, k of quaternion
+
+    Returns
+    -------
+    exp(q) : the quaternion exponential
+    
+    Notes
+    -----
+    See: https://en.wikipedia.org/wiki/Quaternion#Exponential,_logarithm,_and_power
+    '''
+    q = np.array(q) #to ensure there is a dtype
+    w, v = q[0], q[1:]
+    norm = np.sqrt(np.dot(v, v))
+    result = np.zeros((4,), q.dtype)
+
+    if norm == 0.:
```

Comment:
I think you'll need to allow the norm to be close to 1, like : https://github.com/matthew-brett/transforms3d/blob/master/transforms3d/quaternions.py#L93 .  That can be a bit annoying to test, of course.

Reply:
I may not be fully awake yet, but I don't see the parallel between both code sections. I'm only trying to prevent division by 0. returning qeye is questionnable, what matlab does?

While we're at it, returning qeye when qnorm_ is zero is also questionable, I could always let python choke on it.

Reply:
I think Matlab does just do something like what you're doing:

```
>> quatexp([1, 0, 0, 0])

ans =

     1     0     0     0

>> quatexp([10, 0, 0, 0])

ans =

     1     0     0     0

>> quatexp([10, 0, 0, eps])

ans =

   1.0e+04 *

    2.2026         0         0    0.0000

>> quatexp([10, 0, 0, eps / 10])

ans =

   1.0e+04 *

    2.2026         0         0    0.0000

>> quatexp([10, 0, 0, 1e-16])   

ans =

   1.0e+04 *

    2.2026         0         0    0.0000

>> quatexp([10, 0, 0, 1e-18])

ans =

   1.0e+04 *

    2.2026         0         0    0.0000

>> quatexp([10, 0, 0, 1e-20])

ans =

   1.0e+04 *

    2.2026         0         0    0.0000

>> 1e-20

ans =

   1.0000e-20

>> quatexp([10, 0, 0, 1e-100])

ans =

   1.0e+04 *

    2.2026         0         0    0.0000

>> quatexp([10, 0, 0, 1e-1000])

ans =

     1     0     0     0

>> quatexp([10, 0, 0, 1e-106]) 

ans =

   1.0e+04 *

    2.2026         0         0    0.0000
```

Reply:
So, what do you want me to do? I still don't understand what you mean by  
> I think you'll need to allow the norm to be close to 1

Reply:
Sorry - I meant - yes that's fine not to check for a norm close to 1, but just do the obvious thing, as you're doing.

Reply:
Let me know if you have other issues you want me to fix.

Thanks again!

Hope it makes it for 0.3.2 :)

Reply:
Hi matthew, 

Do you plan to pull anytime soon?

these guys are useful for closed form motion integrals...

`
                   current_v, current_w = ...

                    w_pure = np.array([0., current_w[0], current_w[1], current_w[2]], np.float32)

                    # simple kinematic model:
                    p += rotate_vector(current_v, q) * intergation_dt
                    q = qmult(qexp(w_pure * intergation_dt/2.0), q)
                    q /= np.sqrt(qnorm(q))
`

Reply:
I'm really sorry, I got side-tracked making some test data with Matlab.

That's now in the master branch in `transforms3d/tests/data/processed_quats.mat` - see `transforms3d/tests/data/process_quats.m`.  I was going to then write tests of your functions, confirming they get the same result as those generated by Matlab.

Do you have 30 minutes or so to do that?  Otherwise I will get to it fairly soon.

Reply:
I don't have access to a matlab license, can-I use Octave?

Reply:
No - sorry - I meant this : https://github.com/MaximeLemonnier-Leddartech/transforms3d/pull/1

Do these look right to you?